lib/gc/Dialect/Linalgx/Utils.cpp

```diff
@@ -0,0 +1,451 @@
+      }
+    }
+    return matchBatch;
+  };
+  auto matchDims = [](ArrayRef<PackingMap> mapArray, ShapedType firstShape,
+                      ShapedType secondShape) {
+    for (auto &packingMap : mapArray) {
+      bool isDynamic = false;
+      int64_t firstSize = 1;
+      auto firstDims = packingMap.getFirst();
+      for (auto dim : firstDims) {
+        auto size = firstShape.getDimSize(dim);
+        if (size == ShapedType::kDynamic)
+          isDynamic = true;
+        firstSize *= size;
+      }
+      int64_t secondSize = 1;
+      auto secondDims = packingMap.getSecond();
+      for (auto dim : secondDims) {
+        auto size = secondShape.getDimSize(dim);
+        if (size == ShapedType::kDynamic)
+          isDynamic = true;
+        secondSize *= size;
+      }
+      if (isDynamic)
+        return false; // does not support dynamic dims
+      if (firstSize != secondSize)
+        return false;
+    }
+    return true;
+  };
+  bool matchM = matchDims(attr.mPacking, shapeA, shapeC);
+  bool matchN = matchDims(attr.nPacking, shapeB, shapeC);
+  bool matchK = matchDims(attr.kPacking, shapeA, shapeB);
+  bool checkMatch = matchBatch(attr.batchDimMap) && matchM && matchN && matchK;
+  if (!checkMatch)
+    return false;
+
+  return true;
+}
+
+//****************************************************************************//
+//                      IteratorTypes Utils                                   //
+//****************************************************************************//
+
+SmallVector<utils::IteratorType>
+getIteratorTypesArray(const PackingAttr &attr) {
+  SmallVector<utils::IteratorType> iteratorTypes;
+  // get packing num for each packing map
+  auto getBatchIteratorTypes = [&](const BatchDimMap &batchDimMap) {
+    iteratorTypes.insert(iteratorTypes.end(), batchDimMap.getBatchNum(),
+                         batchDimMap.isBatchReduce()
+                             ? utils::IteratorType::reduction
+                             : utils::IteratorType::parallel);
+  };
+  auto getPackingIteratorTypes = [&](ArrayRef<PackingMap> packingMaps,
+                                     utils::IteratorType iterTy) {
+    for (auto &mapping : packingMaps) {
+      auto packingNum = mapping.getPackingDstDims().size();
+      iteratorTypes.insert(iteratorTypes.end(), packingNum, iterTy);
+    }
+  };
+  // Process order: b, m, n, k packing
+  getBatchIteratorTypes(attr.batchDimMap);
+  getPackingIteratorTypes(attr.mPacking, utils::IteratorType::parallel);
+  getPackingIteratorTypes(attr.nPacking, utils::IteratorType::parallel);
+  getPackingIteratorTypes(attr.kPacking, utils::IteratorType::reduction);
+  return iteratorTypes;
+}
+
+//****************************************************************************//
+//                     IndexingMaps Utils                                     //
+//****************************************************************************//
+
```

Comment:
DO we need to align conduct of coding with upstream LLVM? I.e. use `///` commenter instead of `// ***`